### ISSUE TITLE
beforeRouteUpdate not triggered but component is reused

### ISSUE THREAD

**Issue by matzeso** [vuejs/vue-router#3068]
### Version
3.1.3

### Reproduction link
[https://codesandbox.io/s/vue-router-update-keep-alive-lgi16](https://codesandbox.io/s/vue-router-update-keep-alive-lgi16)

### Steps to reproduce
Click on dashboard
-> click item 1

`created()`/ `mounted()` hooks are fired and are setting some variable based on query param

-> click _Go back_ or _Dashboard_

-> click _Item 2_

created() or mounted() are not fired again, because the `router-view` is wrapped within `keep-alive`.
Therefore, the component already exists and will not be re-created.

But, since the route changes, from `dashboard` to `item`, `beforeRouteEnter` is called instead of `beforeRouteUpdate`. But `beforeRouteEnter` can not modify existing state (or even access it). 

### What is expected?
Since the component is reused, I would expect `beforeRouteUpdate` to be fired.

In my situation, if my query param changes, I need to fetch new data from the backend and store them in local state. As of now, there is no easy way to detect that my param changes.

I would need to add a watcher on $route, but that watcher will run in every other child component since my component is not destroyed. 

### What is actually happening?
`beforeRouteEnter` is fired

---
With this situation, if I keep using `keep-alive` , it makes it pretty hard to detect when my page is openend again. 
Since my watchers would still run even when I am visiting other child-pages, there is a lot of overhead of introducing a $route watcher that performs some action when my query params change. In addition, I would need to make sure that the action only triggers when I am actually seeing this sub-page (requiring my component to be aware of its route, which feels wrong). 

I think that in cases with existing components, Vue Router should fire `beforeRouteUpdate` even when the route changes. 
I do understand though that technically the route is changing so it makes sense to call beforeRouteEnter instead of Update. 
But then, it would help if the second beforeRouteEnter, that is called when then component already exists, has access to `this`.

**Comment by LinusBorg** [comment]
> I do understand though that technically the route is changing so it makes sense to call beforeRouteEnter instead of Update.

I agree. Semantically, `Enter` is the corret hook.

> But then, it would help if the second beforeRouteEnter, that is called when then component already exists, has access to `this`.

I tend to agree but am unsure wether this is easily doable in the current architecture. @posva might be able to judge this better.

> Since my watchers would still run even when I am visiting other child-pages, there is a lot of overhead of introducing a $route watcher that performs some action when my query params change.

I don't agree that there's noteworthy overhead. For one, rout events are user-triggered events and don't fire in rapid succession.

Furthermore, each watcher can check wether the new route is thew one it's supposed to be triggered by, and if it's not, just do nothing. That's a one-liner and as cheap ans any op can be.

Alternative, you could use keep-alive's `activated` and `deactivated` lifecycle hooks to manage an internal flag which the watcher can use to know wether the component is active and therefore, wether it should run.

**Comment by posva** [comment]
> But then, it would help if the second beforeRouteEnter, that is called when then component already exists, has access to this.

I don't think making the hook behave differently based on whether the component is reused or not, is a good thing because it's inconsistent and would lead to confusing behavior.

In addition to the solution given by @LinusBorg by using a flag, you could also stop the watchers on `deactivated` and read them on `activated`

Router hooks depend on the route being reused, they are not dependent of the components being reused. This is intentional, as you could use component lifecycle hooks instead otherwise

**Comment by matthias-influry** [comment]
Hey both of you, thanks for your answers.
I did not know about `activated` and `deactivated`. 
I was missing exactly those hooks. Thanks for pointing that out. 
I looked through the lifecycle hook graph but they were not there, should have checked the API docs as well. 

> Router hooks depend on the route being reused, they are not dependent of the components being reused. This is intentional, as you could use component lifecycle hooks instead otherwise

That makes sense. Especially now that I know there does indeed exist a hook that I can use to check that my component is reused (by being activated again).
